smarts/core/plan.py

```diff
@@ -0,0 +1,332 @@
+
+
+@dataclass(frozen=True)
+class VehicleSpec:
+    veh_id: str
+    veh_config_type: str
+    dimensions: Dimensions
+
+
+@dataclass(frozen=True)
+class Mission:
+    start: Start
+    goal: Goal
+    # An optional list of road IDs between the start and end goal that we want to
+    # ensure the mission includes
+    route_vias: Tuple[str] = field(default_factory=tuple)
+    start_time: float = 0.1
+    entry_tactic: EntryTactic = None
+    via: Tuple[Via, ...] = ()
+    # if specified, will use vehicle_spec to build the vehicle (for histories)
+    vehicle_spec: VehicleSpec = None
+
+    @property
+    def has_fixed_route(self) -> bool:
+        return not self.goal.is_endless()
+
+    def is_complete(self, vehicle, distance_travelled: float) -> bool:
+        return self.goal.is_reached(vehicle)
+
+    @staticmethod
+    def random_endless_mission(
+        road_map: RoadMap,
+        min_range_along_lane: float = 0.3,
+        max_range_along_lane: float = 0.9,
+    ) -> Mission:
+        assert min_range_along_lane > 0  # Need to start further than beginning of lane
+        assert max_range_along_lane < 1  # Cannot start past end of lane
+        assert min_range_along_lane < max_range_along_lane  # Min must be less than max
+
+        road = road_map.random_route(1).roads[0]
+        n_lane = random.choice(road.lanes)
+
+        # XXX: The ends of the road are not as useful as starting mission locations.
+        #      Sumo complains if we get too close to 0 or `lane_length`.
+        offset = random.random() * min_range_along_lane + (
+            max_range_along_lane - min_range_along_lane
+        )
+        offset *= n_lane.length
+        coord = n_lane.from_lane_coord(RefLinePoint(offset))
+        target_pose = n_lane.target_pose_at_point(coord)
+        return Mission(
+            start=Start(target_pose.position, target_pose.heading),
+            goal=EndlessGoal(),
+            entry_tactic=None,
+        )
+
+
+@dataclass(frozen=True)
+class LapMission:
+    start: Start
+    goal: Goal
+    route_length: float
+    num_laps: int = None  # None means infinite # of laps
+    # An optional list of road IDs between the start and end goal that we want to
+    # ensure the mission includes
+    route_vias: Tuple[str] = field(default_factory=tuple)
+    start_time: float = 0.1
+    entry_tactic: EntryTactic = None
+    via_points: Tuple[Via, ...] = ()
+
+    @property
+    def has_fixed_route(self) -> bool:
+        return True
+
+    def is_complete(self, vehicle, distance_travelled: float) -> bool:
+        return (
+            self.goal.is_reached(vehicle)
+            and distance_travelled > self.route_length * self.num_laps
+        )
+
+
+class Plan:
+    def __init__(
+        self, road_map: RoadMap, mission: Mission = None, find_route: bool = True
+    ):
+        self._road_map = road_map
+        self._mission = mission
+        self._route = None
+        if find_route:
+            self.create_route(mission)
```

Comment:
I agree with your sense that this is sub-optimal.  It's currently like this because of one use case, which is in `_prepare_sensors_for_agent_control()` of `bubble_manager.py`.  Otherwise, we could just _always_ have the constructor call `find_route()` and ditch the `find_route` parameter.   When I was refactoring, I didn't look too deeply into how to improve the `_prepare_sensors_for_agent_control()` method, but perhaps there's some better way to do it such that we can make this change.  Any ideas?